go/libraries/events/directory_proc.go

```diff
@@ -0,0 +1,208 @@
+			dp.fileProcessingWorker(fileCh, compCh, stopCh)
+		}()
+	}
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		dp.watchLoop(fileCh, compCh, stopCh)
+	}()
+
+	return dp
+}
+
+// Close closes a DirectoryProcessor and cleans up all of it's resources
+func (proc *DirectoryProcessor) Close() error {
+	close(proc.stopCh)
+	proc.wg.Wait()
+
+	return nil
+}
+
+func (proc *DirectoryProcessor) fileProcessingWorker(fileCh <-chan FileAndSize, compCh chan<- string, stopCh <-chan struct{}) {
+	for {
+		if isStopped(stopCh) {
+			return
+		}
+
+		select {
+		case <-stopCh:
+			return
+
+		case fileAndSize, ok := <-fileCh:
+			if !ok {
+				return
+			}
+
+			f := func() error {
+				defer func() { compCh <- fileAndSize.File }()
+				ctx, _ := context.WithDeadline(context.Background(), time.Now().Add(proc.TimeToProcess))
+				err := proc.ProcessFunc(ctx, proc.FS, fileAndSize.Size, fileAndSize.File)
+
+				if err != nil {
+					return err
+				}
+
+				err = proc.FS.Delete(fileAndSize.File, false)
+
+				if err != nil {
+					return err
+				}
+
+				return nil
+			}
+
+			err := f()
+
+			if err != nil {
+				// TODO: fix logging
+				log.Println(err)
+			}
+		}
+	}
+}
+
+func (proc *DirectoryProcessor) watchLoop(fileCh chan<- FileAndSize, compCh <-chan string, stopCh <-chan struct{}) {
+	defer close(fileCh)
+
+	beingProcessed := make(map[string]struct{})
+
+	for {
+		if isStopped(stopCh) {
+			return
+		}
+
+		nonEmpty := make(map[string]struct{})
+		dirs := make(map[string]struct{})
+		toProcess := make(map[string]int64)
+		err := proc.FS.Iter(proc.Path, true, func(path string, size int64, isDir bool) (stop bool) {
+			if isDir {
+				dirs[path] = struct{}{}
+			} else if filepath.Ext(path) == evtDataExt {
+				nonEmpty[filepath.Dir(path)] = struct{}{}
+				toProcess[path] = size
+			}
+
+			return false
+		})
+
+		if err != nil {
+			// TODO: fix logging
+			log.Printf("error iterating over '%s': %v", proc.Path, err)
+
+			time.Sleep(proc.BetweenIterations)
+			continue
+		}
+
+		for k := range nonEmpty {
+			delete(dirs, k)
+		}
+
+		for k := range dirs {
+			modTime, exists := proc.FS.LastModified(k)
+
+			if exists && time.Since(modTime) > delDirDuration {
+				_ = proc.FS.Delete(k, false)
+			}
+		}
+
+		for k := range toProcess {
+			if _, ok := beingProcessed[k]; ok {
+				delete(toProcess, k)
+			}
+		}
+
+		for k, v := range toProcess {
+			drainAvailableCompleted(compCh, beingProcessed)
+
+			beingProcessed[k] = struct{}{}
+			fileCh <- FileAndSize{File: k, Size: v}
+		}
+
+		time.Sleep(proc.BetweenIterations)
+		drainAvailableCompleted(compCh, beingProcessed)
+	}
+}
+
+func isStopped(stopCh <-chan struct{}) bool {
+	select {
+	case <-stopCh:
+		return true
+
+	default:
+	}
+
+	return false
+}
+
+func drainAvailableCompleted(compCh <-chan string, beingProcessed map[string]struct{}) {
+	empty := false
+	for !empty {
+		select {
+		case completed := <-compCh:
+			delete(beingProcessed, completed)
+
+		default:
+			empty = true
+		}
+	}
+}
```

Comment:
Same code from the events server. Just changed the log extension variable to be `evtDataExt`.
Will add the test file as well.

Reply:
I think this might be a bit of overkill and could result in a ton of send-metrics processes running when someone is working in offline mode.  I think instead of watching the directory and looping as long and there are still files, you should just loop over all files until you have a failure to send, and then exit on first failure.